README
NAME
    RT::Extension::ReferenceIDoitObjects - Create a ticket in relation to
    one or more i-doit objects

DESCRIPTION
    This extension gives you the opportunity to combine an issue tracker
    like RT with an IT documentation tool / CMDB like i-doit. It uses
    i-doit's API to relate a ticket with one or more CIs / objects managed
    by i-doit. On i-doit's side you are able to view all tickets related to
    an object. This extension also supports i-doit's multi-mandator
    functionality.

    i-doit ("I document IT") is a web-based tool to document complex IT
    infrastructures. It provides several modules like a ITIL compliant
    Configuration Management Database (CMDB). More information about i-doit
    is available under <http://www.i-doit.org/>. Its core is Free and Open
    Source Software. Visit <http://www.i-doit.com/> for commercial support
    and additional services.

INSTALLATION
    This extension requires RT 4.2.x or higher and a running installation of
    i-doit 1.4 or higher.

    To install this extension, run the following commands:

        perl Makefile.PL
        make
        make test
        make install
        make initdb

    Executing the last command creates 2 new custom fields, so please do it
    only once. These fields contain the i-doit mandator and the referenced
    objects.

    Another way to install the latest release is via CPAN:

        cpan RT::Extension::ReferenceIDoitObjects
        $RT_HOME/sbin/rt-setup-database --action insert --datafile /opt/rt4/local/plugins/RT-Extension-ReferenceIDoitObjects/etc/initialdata

    The second command is equivalent to "make initdb", but is unfortunately
    not executed automatically. $RT_HOME is the path to your RT
    installation, for example "/opt/rt4".

CONFIGURATION
    To enable this extension edit the RT site configuration based in
    "$RT_HOME/etc/RT_SiteConfig.pm":

        Set(@Plugins,qw(RT::Extension::ReferenceIDoitObjects));

    i-doit has a built-in API based on JSON-RPC. To call this API set its
    URL:

        Set($IDoitURL, 'http://example.org/i-doit/');

        Set($IDoitAPI, $IDoitURL . '?api=jsonrpc');

        Set(%IDoitMandatorKeys, (
            'Mandator 1' => 'api key',
            'Mandator 2' => 'api key'
        ));

        Set($IDoitDefaultMandator, 'Mandator 1');

        Set($IDoitDefaultView, 'objects'); # 'objects', 'workplaces', 'devices', or 'item'

        Set($IDoitInstalledSoftware, 'relations'); # 'objects', or 'relations'

        Set($IDoitShowCustomFields, 1); # 1 ('yes') or 0 ('no')

    $IDoitURL
        It's *highly recommended* to establish an encrypted connection
        between RT and i-doit over a network, e. g. TLS over HTTP (HTTPS).

    $IDoitAPI
        Please be aware of browsers' "Same Origin Policy"! This extension
        uses AJAX requests access i-doit's API. If RT and i-doit are not
        available under the same domain name (or IP address), AJAX calls
        will fail.

        To avoid this "problem" (actually this policy is very useful) you
        can setup an AJAX proxy. This extension already provides such a
        proxy located under "etc/". It's written in PHP, so you have to
        install PHP 5.2 or higher and the PHP extension "curl" on the same
        machine where RT is installed. Make this little script available
        through your web server and edit the script by setting $l_url to the
        URL of i-doit's API, e. g.
        "http://example.org/i-doit/index.php?api=jsonrpc". In RT's site
        configuration $IDoitAPI has to be set to this script, e. g.
        "http://rt.example.org/path/to/i-doit_api_proxy.php".

    $IDoitMandatorKeys
        This is a list of mandators with their API keys. Just put the name
        and API key of every mandator in i-doit you like to relate to
        tickets.

        Notice: Within the Web GUI you must configure the custom field
        "i-doit mandator". Add a new value for each mandator. The important
        field is "name" where you should set the mandator name.

    $IDoitDefaultMandator
        Choose a default mandator for every situation where it's needed. Use
        its name (or whatever you like) to identify the mandator. This name
        has be to added to the list of the corresponding custom field as
        well.

    $IDoitDefaultView
        When creating or editing a ticket, this extension adds a so-called
        *object browser* to the web interface. The browser gives you several
        views on objects:

        "objects"
            Select objects provided by the API and filter them by type.

        "workplaces"
            Select users' workplaces and their related sub objects. Each
            user will be taken by the email address provided by RT's field
            "Requestors" if these users are documented in i-doit.

            i-doit gives you the possiblity to create relations between
            users, their workplaces and all equipment related to these
            workplaces.

            Tip: You may synchronize user information between OTRS and
            i-doit via LDAP.

        "devices"
            Select assigned devices for current requestor. Those devices are
            objects in i-doit which have this requestor as an assigend
            person.

        "selected"
            View and remove all selected items.

    $IDoitInstalledSoftware
        Defines which type of objects will be shown for the installed
        software. There are two options: "objects" or "relations".

        "objects"
            Shows software objects which are assigned to the currently
            selected object.

        "relations"
            Shows the software relation between the object and the assigned
            software.

    $IDoitShowCustomFields
        Sometimes it's better to "clean up" the web user interface. Whenever
        you only have 1 mandator within i-doit and don't want to edit the
        object identifiers manually it's recommended to hide the used custom
        fields. Select 1 to show them or 0 to hide them.

    After all your new configuration will take effect after restarting your
    RT environment:

        rm -rf $RT_HOME/var/mason_data/obj/* && service apache2 restart

    This is an example for deleting the mason cache and restarting the
    Apache HTTP web server on a Debian GNU/Linux based operating system.

I-DOIT CONFIGURATION
    You may see and create object-related tickets within i-doit. Please
    refer to the i-doit manual to enable this feature.

    If you create a new ticket in i-doit a new browser tab will be opened
    with the RT user interface. Sometimes RT shows a warning that there is a
    CSR attack. If you observe this behavior edit RT's local configuration
    file "$RT_HOME/etc/RT_SiteConfig.pm" where $RT_HOME is the path to your
    RT installation, for example "/opt/rt4":

        Set($RestrictReferrer, 0); # avoids possible CSR attacks

    Don't forget to clear the Mason cache and restart your webserver.

    Notice: This setting could breach your security!

USAGE
    Whenever you create a new ticket or edit an existing one you are able to
    reference this ticket with one or more objects in i-doit. An additional
    box with the so-called "object browser" will shown up. Just select the
    objects you need or unselect the objects you don't need.

AUTHORS
    Benjamin Heisig, <bheisig@synetics.de>

    Leonard Fischer, <lfischer@synetics.de>

    Van Quyen Hoang, <qhoang@synetics.de>

SUPPORT AND DOCUMENTATION
    You can find documentation for this module with the "perldoc" command.

        perldoc RT::Extension::ReferenceIDoitObjects

    You can also look for information at:

    Search CPAN
        <http://search.cpan.org/dist/RT-Extension-ReferenceIDoitObjects/>

    RT: CPAN's request tracker
        <http://rt.cpan.org/NoAuth/Bugs.html?Dist=RT-Extension-ReferenceIDoi
        tObjects>

    AnnoCPAN: Annotated CPAN documentation
        <http://annocpan.org/dist/RT-Extension-ReferenceIDoitObjects>

    CPAN Ratings
        <http://cpanratings.perl.org/d/RT-Extension-ReferenceIDoitObjects>

    Repository
        <https://github.com/bheisig/rt-extension-referenceidoitobjects>

BUGS
    Please report any bugs or feature requests to the authors.

COPYRIGHT AND LICENSE
    Copyright (C) 2011-14 synetics GmbH, <http://i-doit.org/>

    This program is free software: you can redistribute it and/or modify it
    under the terms of the GNU Affero General Public License as published by
    the Free Software Foundation, either version 3 of the License, or (at
    your option) any later version.

    This program is distributed in the hope that it will be useful, but
    WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero
    General Public License for more details.

    You should have received a copy of the GNU Affero General Public License
    along with this program. If not, see <http://www.gnu.org/licenses/>.

    Request Tracker (RT) is Copyright Best Practical Solutions, LLC.

SEE ALSO
        RT



Changes
#   Changelog

Version history for RT-Extension-ReferenceIDoitObjects


##  1.00 (tba)

**Important notice:** This extension requires Request Tracker 4.4.x and i-doit 1.8.2 or higher. It is neither compatible to RT 4.0, 4.2 nor i-doit <= 1.8.1.

*   Fix: Re-named "mandator" to "tenant"
*   Fix: Re-written documentation
*   Fix: Clarify license (AGPLv3)
*   Fix: Broken include of Elements/EditCustomFields
*   Fix: Removed legacy IE hack


##  0.94 (Fri, 19 Sep 2014 11:16:42 +0200)

This is a bugfix-only release.

*   Fix: Improved selecting object type
*   Fix: Fixed missing tab navigation after submitting ticket changes on referenced i-doit objects
*   Fix: Fetching devices related to current requestor doesn't work
*   Fix: Fetching right object type when selecting an object
*   Fix: (Un)selecting all object items in a table
*   Fix: Selecting objects in all tables at the same time
*   Fix: Improved table styling
*   Fix: Broken images
*   Fix: Removed non-relevant columns from devices table
*   Fix: Improved sorting table columns
*   Fix: Avoid CSR error message when reloading "create new ticket" page
*   Fix: Show configured default view when initiated
*   Fix: Creating new logbook entries
*   Fix: Updated dataTables jQuery plugin
*   Fix: Re-factored JavaScript code
*   Fix: Improved code documentation


##  0.93 (Wed, 07 May 2014 14:20:10 +0200)

**Important notice:** This extension requires Request Tracker 4.2.x or higher and a running installation of i-doit 1.3 or higher. It is no longer compatible to RT version 4.0.x and i-doit version 1.2.x or older! Please use version 0.92 instead if you need compatibility to those older versions.

*   Made the code compatible to Request Tracker 4.2.x (breaks with 4.0.x)
*   Switched from username/password authentication to i-doit API key
*   Added view for requestor's linked devices including installed software components
*   Renamed some tabs for better understanding
*   Shared code with OTRS Help Desk extension to be more maintainable
*   Renamed configuration options for default view


##  0.92 (Fri, 07 Jun 2013 14:03:02 +0200)

*   Object browser shows object location
*   Implemented browser in "Jumbo update"
*   Hide custom fields when modifying referenced objects
*   Tree view: mark both hard- and software if software is marked
*   Fixed broken link in page menu for /Ticket/ModifyIDoitObjects.html after update
*   Removed buggy unused code
*   Improved code documentation
*   Updated to Module::Install 1.06 (thanks to Tatsuhiko Miyagawa)
*   Requires i-doit version 1.0.1 or higher


##  0.91 (Fri, 07 Sep 2012 14:02:49 +0200)

*   Removed deprecated Switch.pm (important for newer Perl versions; thanks to Matthias)
*   Described installation via CPAN.


##  0.9 (Mon, 04 Jun 2012 16:50:31 +0200)

*   Include javascript and css only if needed
*   Added checkbox to select/unselect all objects from object view
*   Added link to remove all selected objects from list
*   Show message if no requestor is selected
*   Implemented workaround for JavaScript's JSON object missing in IE < 9


##  0.8 (Fri, 01 Jun 2012 14:12:24 +0200)

*   Show link in the page menu to modify referenced i-doit objects


##  0.7 (Thu, 31 May 2012 10:01:16 +0200)

*   Use default tenant if necessary when modifying referenced i-doit objects
*   Look only for enabled object types (bug fix)


##  0.6 (Fri, 27 Apr 2012 11:49:06 +0200)

*   Write logbook entries within i-doit when objects are added or removed
*   Show loading message when initiating the browser
*   Requires i-doit version 0.9.9-9 or higher


##  0.5 (Mon, 23 Apr 2012 13:22:37 +0200)

*   When creating a ticket assign tenant and objects from HTTP GET
*   Bug fixing


##  0.4 (skipped)


##  0.3 (Mon, 13 Feb 2012 13:28:50 +0100)

*   Performance boost when rendering tables
*   Fixed API call when viewing a ticket created by email
*   Improved error messages
*   Improved meta information


##  0.2 (Fri, 20 Jan 2012 15:45:43 +0100)

*   Bug fixing


##  0.1 (Wed, 14 Dec 2011 09:30:22 +0100)

*   Initial release



lib/RT/Extension/ReferenceIDoitObjects.pm
# RT::Extension::ReferenceIDoitObjects
#
# Copyright (C) 2011-17 synetics GmbH, <https://i-doit.com/>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as
# published by the Free Software Foundation, either version 3 of the
# License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
# Request Tracker (RT) is Copyright Best Practical Solutions, LLC.

package RT::Extension::ReferenceIDoitObjects;

use 5.010;
use warnings;
use strict;

our $VERSION = '1.00';

=head1 NAME

RT::Extension::ReferenceIDoitObjects - Create a ticket in relation to one or more i-doit objects


=head1 DESCRIPTION

This extension gives you the opportunity to combine an issue tracker like RT with an IT documentation tool / CMDB like i-doit. It uses i-doit's API to relate a ticket with one or more CIs / objects managed by i-doit. On i-doit's side you are able to view all tickets related to an object. This extension also supports i-doit's multi-tenant functionality.

i-doit ("I document IT") is a web-based tool to document complex IT infrastructures. It provides several modules like a ITIL compliant Configuration Management Database (CMDB). More information about i-doit is available under L<http://www.i-doit.org/>. Its core is Free and Open Source Software. Visit L<http://www.i-doit.com/> for commercial support and additional services.


=head1 INSTALLATION

This extension requires RT 4.2.x or higher and a running installation of i-doit 1.4 or higher.

To install this extension, run the following commands:

    perl Makefile.PL
    make
    make test
    make install
    make initdb

Executing the last command creates 2 new custom fields, so please do it only once. These fields contain the i-doit tenant and the referenced objects.

Another way to install the latest release is via CPAN:

    cpan RT::Extension::ReferenceIDoitObjects
    $RT_HOME/sbin/rt-setup-database --action insert --datafile /opt/rt4/local/plugins/RT-Extension-ReferenceIDoitObjects/etc/initialdata

The second command is equivalent to C<make initdb>, but is unfortunately not executed automatically. C<$RT_HOME> is the path to your RT installation, for example C</opt/rt4>.


=head1 CONFIGURATION

To enable this extension edit the RT site configuration based in C<$RT_HOME/etc/RT_SiteConfig.pm>:

    Set(@Plugins,qw(RT::Extension::ReferenceIDoitObjects));

i-doit has a built-in API based on JSON-RPC. To call this API set its URL:

    Set($IDoitURL, 'http://example.org/i-doit/');

    Set($IDoitAPI, $IDoitURL . '?api=jsonrpc');

    Set(%IDoitTenantKeys, (
        'Tenant 1' => 'api key',
        'Tenant 2' => 'api key'
    ));

    Set($IDoitDefaultTenant, 'Tenant 1');

    Set($IDoitDefaultView, 'objects'); # 'objects', 'workplaces', 'devices', or 'item'

    Set($IDoitInstalledSoftware, 'relations'); # 'objects', or 'relations'

    Set($IDoitShowCustomFields, 1); # 1 ('yes') or 0 ('no')


=over 4

=item C<$IDoitURL>

It's I<highly recommended> to establish an encrypted connection between RT and i-doit over a network, e. g. TLS over HTTP (HTTPS).


=item C<$IDoitAPI>

Please be aware of browsers' "B<Same Origin Policy>"! This extension uses AJAX requests access i-doit's API. If RT and i-doit are not available under the same domain name (or IP address), AJAX calls will fail.

To avoid this "problem" (actually this policy is very useful) you can setup an AJAX proxy. This extension already provides such a proxy located under C<etc/>. It's written in PHP, so you have to install PHP 5.2 or higher and the PHP extension C<curl> on the same machine where RT is installed. Make this little script available through your web server and edit the script by setting C<$l_url> to the URL of i-doit's API, e. g. C<http://example.org/i-doit/index.php?api=jsonrpc>. In RT's site configuration C<$IDoitAPI> has to be set to this script, e. g. C<http://rt.example.org/path/to/i-doit_api_proxy.php>.


=item C<$IDoitTenantKeys>

This is a list of tenants with their API keys. Just put the name and API key of every tenant in i-doit you like to relate to tickets.

B<Notice:> Within the Web GUI you must configure the custom field "i-doit tenant". Add a new value for each tenant. The important field is C<name> where you should set the tenant name.


=item C<$IDoitDefaultTenant>

Choose a default tenant for every situation where it's needed. Use its name (or whatever you like) to identify the tenant. This name has be to added to the list of the corresponding custom field as well.


=item C<$IDoitDefaultView>

When creating or editing a ticket, this extension adds a so-called I<object browser> to the web interface. The browser gives you several views on objects:

=over 4

=item C<objects>

Select objects provided by the API and filter them by type.

=item C<workplaces>

Select users' workplaces and their related sub objects. Each user will be taken by the email address provided by RT's field "Requestors" if these users are documented in i-doit.

i-doit gives you the possiblity to create relations between users, their workplaces and all equipment related to these workplaces.

Tip: You may synchronize user information between OTRS and i-doit via LDAP.

=item C<devices>

Select assigned devices for current requestor. Those devices are objects in i-doit which have this requestor as an assigend person.

=item C<selected>

View and remove all selected items.

=back


=item C<$IDoitInstalledSoftware>

Defines which type of objects will be shown for the installed software. There are two options: "objects" or "relations".


=over 4

=item C<objects>

Shows software objects which are assigned to the currently selected object.

=item C<relations>

Shows the software relation between the object and the assigned software.

=back


=item C<$IDoitShowCustomFields>

Sometimes it's better to "clean up" the web user interface. Whenever you only have 1 tenant within i-doit and don't want to edit the object identifiers manually it's recommended to hide the used custom fields. Select 1 to show them or 0 to hide them.


=back


After all your new configuration will take effect after restarting your RT environment:

    rm -rf $RT_HOME/var/mason_data/obj/* && service apache2 restart

This is an example for deleting the mason cache and restarting the Apache HTTP web server on a Debian GNU/Linux based operating system.


=head1 I-DOIT CONFIGURATION

You may see and create object-related tickets within i-doit. Please refer to the i-doit manual to enable this feature.

If you create a new ticket in i-doit a new browser tab will be opened with the RT user interface. Sometimes RT shows a warning that there is a CSR attack. If you observe this behavior edit RT's local configuration file C<$RT_HOME/etc/RT_SiteConfig.pm> where C<$RT_HOME> is the path to your RT installation, for example C</opt/rt4>:

    Set($RestrictReferrer, 0); # avoids possible CSR attacks

Don't forget to clear the Mason cache and restart your webserver.

B<Notice:> This setting could breach your security!


=head1 USAGE

Whenever you create a new ticket or edit an existing one you are able to reference this ticket with one or more objects in i-doit. An additional box with the so-called "object browser" will shown up. Just select the objects you need or unselect the objects you don't need.


=head1 AUTHORS

Benjamin Heisig, E<lt>bheisig@synetics.deE<gt>

Leonard Fischer, E<lt>lfischer@synetics.deE<gt>

Van Quyen Hoang, E<lt>qhoang@synetics.deE<gt>


=head1 SUPPORT AND DOCUMENTATION

You can find documentation for this module with the C<perldoc> command.

    perldoc RT::Extension::ReferenceIDoitObjects

You can also look for information at:


=over 4

=item B<Search CPAN>

L<http://search.cpan.org/dist/RT-Extension-ReferenceIDoitObjects/>

=item B<RT: CPAN's request tracker>

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=RT-Extension-ReferenceIDoitObjects>

=item B<AnnoCPAN: Annotated CPAN documentation>

L<http://annocpan.org/dist/RT-Extension-ReferenceIDoitObjects>

=item B<CPAN Ratings>

L<http://cpanratings.perl.org/d/RT-Extension-ReferenceIDoitObjects>

=item B<Repository>

L<https://github.com/bheisig/rt-extension-referenceidoitobjects>

=back


=head1 BUGS

Please report any bugs or feature requests to the L<authors|/"AUTHORS">.


=head1 COPYRIGHT AND LICENSE

Copyright (C) 2011-17 synetics GmbH, E<lt>https://i-doit.com/E<gt>

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU Affero General Public License as
published by the Free Software Foundation, either version 3 of the
License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Affero General Public License for more details.

You should have received a copy of the GNU Affero General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.

Request Tracker (RT) is Copyright Best Practical Solutions, LLC.


=head1 SEE ALSO

    RT


=cut

1;
